### PR TITLE
feat(api, robot-server): allow setting CSV param files via protocols & analyses endpoints

### PR DESCRIPTION
**Pull request by sanni-t**
Closes AUTH-416

# Overview

This PR adds the capability of setting the CSV run time parameter with a CSV file. Features added:
- `POST /protocols` and `POST /analyses` now accept `runTimeParameterFiles` data in addition to the existing `runTimeParameterValues` data
- Adds a table to save the CSV file IDs used in analyses
- The CSV param is updated with the file ID mentioned 

# Test Plan

- Upload the protocol in #15332 without supplying a csv file ID in the request. Test that:
    - [ ] the protocol upload response contains the full run time parameters, including CSV parameter without any file info
    - [ ] the protocol's analysis eventually completes with a status of `ok` and the run time parameters are present in the full completed analysis
- Upload a CSV file to the `/dataFiles` endpoint, get its ID.
- Upload the above protocol again, and provide the above CSV file ID with it using `runTimeParameterFiles`. Test that:
    - [ ] the protocol upload response contains the full run time parameters, including CSV parameter *and the file ID*
    - [ ] the protocol's analysis eventually completes with a status of `ok` and the CSV parameters with file ID are present in the full analysis
- Test both points (1) & (3) using the `POST /protocols/protocolId/analyses` endpoint
- Test that if you send the same protocol with the same RTP values and files, a new analysis is not started

# Changelog

API:
- updated the CSV param engine type to include file name in `file` property
- updated `PythonProtocolRunner`.load & supporting functions to accept CSV file ID overrides and use them to validate and set CSV parameters.
- added a function to initialize the CSV parameter with the file and file info (right now only sets the file ID)

Robot-server:
- added analysis csv params DB table
- updated `POST /protocols` and `POST /analyses` to accept `runTimeParameterFiles` in their form-data & request body, respectively.
- updated analysis stores to update the csv table with CSV file IDs

# Review requests

- Test with the app and make sure the app requirements are fulfilled.
- Usual code sanity checks.

# Risk assessment

None- low. Adds a new capability to the protocols endpoints and doesn't change existing code much.

### CODE REVIEW

File: api/src/opentrons/protocol_engine/types.py [hunk 1]
```diff
@@ -1044,15 +1055,17 @@ class CSVParameter(RTPBase):
...
         default="csv_file", description="String specifying the type of this parameter"
     )
-    file: Optional[FileId] = Field(
-        ...,
-        description="The CSV file stored on the robot, to be used as the CSV RTP override value."
-        " For local analysis this will be empty.",
+    file: Optional[FileInfo] = Field(
+        default=None,
+        description="ID of the CSV file stored on the robot; to be used for fetching the CSV file."
+        " For local analysis this will most likely be empty.",
```
Comment (sanni-t) [inline api/src/opentrons/protocol_engine/types.py:1061]: @ncdiehl @koji : reverted this to be a nested `file` property and added `name` field to the data type.

File: robot-server/tests/integration/http_api/protocols/test_analyses_with_csv_file_parameters.tavern.yaml [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                      value: flex_8channel_50
+                  default: flex_1channel_50
+                  value: flex_1channel_50
+                  description: What pipette to use during the protocol.
+                - displayName: Liquid handling CSV file
+                  variableName: liq_handling_csv_file
+                  description: A CSV file that contains wells to use for pipetting
+                  file:
+                  - id: '{csv_file_id}'
+                    name: ''
```
Comment (sanni-t) [inline robot-server/tests/integration/http_api/protocols/test_analyses_with_csv_file_parameters.tavern.yaml]: @ncdiehl @koji see this for example CSV param in analysis response

File: api/src/opentrons/protocol_api/_parameter_context.py [hunk 3]
```diff
@@ -12,10 +12,15 @@
...
 )
-from opentrons.protocols.parameters.exceptions import ParameterDefinitionError
+from opentrons.protocols.parameters.exceptions import (
+    ParameterDefinitionError,
+    ParameterValueError,
+)
 from opentrons.protocol_engine.types import (
     RunTimeParameter,
     PrimitiveRunTimeParamValuesType,
+    CsvRunTimeParamFilesType,
```
Comment (jbleon95) [inline api/src/opentrons/protocol_api/_parameter_context.py]: Little nitpick, I think we should always capitalize `CSV` if we're doing camelcase, it makes it stand out more.

File: robot-server/robot_server/protocols/completed_analysis_store.py [hunk 4]
```diff
@@ -285,10 +286,30 @@ def get_primitive_rtps_by_analysis_id(
...
+        """Get the saved CSV RTP file IDs from database."""
+        statement = (
+            sqlalchemy.select(analysis_csv_rtp_table)
+            .where(analysis_csv_rtp_table.c.analysis_id == analysis_id)
+            .order_by(sqlite_rowid)
+        )
+        with self._sql_engine.begin() as transaction:
+            results = transaction.execute(statement).all()
+
+        csv_rtps: Dict[str, Union[str, None]] = {}
```
Comment (jbleon95) [inline robot-server/robot_server/protocols/completed_analysis_store.py]:
```suggestion
        csv_rtps: Dict[str, Optional[str]] = {}
```

File: robot-server/robot_server/protocols/completed_analysis_store.py [hunk 5]
```diff
@@ -320,15 +344,23 @@ async def make_room_and_add(
             await completed_analysis_resource.to_sql_values()
         )
         insert_rtp_statement = analysis_primitive_type_rtp_table.insert()
+        insert_csv_rtp_statement = analysis_csv_rtp_table.insert()
 
         with self._sql_engine.begin() as transaction:
             transaction.execute(delete_primitive_rtp_statement)
+            transaction.execute(delete_csv_rtp_statement)
+            transaction.execute(delete_csv_rtp_statement)
```
Comment (jbleon95) [inline robot-server/robot_server/protocols/completed_analysis_store.py]: Line repeated, unless we want to make sure the file is really deleted 😛

File: robot-server/robot_server/protocols/rtp_resources.py [hunk 6]
```diff
@@ -66,3 +66,41 @@ def _int_converter(value: str) -> float:
...
         )
+
+
+@dataclass
+class CsvParameterResource:
+    """A CSV runtime parameter from a completed analysis, storable in a SQL database."""
+
+    analysis_id: str
+    parameter_variable_name: str
+    file_id: Union[str, None]
```
Comment (jbleon95) [inline robot-server/robot_server/protocols/rtp_resources.py]:
```suggestion
    file_id: Optional[str]
```

**Review by jbleon95 — COMMENTED** [review]
Some minor nitpicks with style and the like, but this is looking good!

File: robot-server/robot_server/protocols/completed_analysis_store.py (hunk 5 above)
Comment (sanni-t) [reply robot-server/robot_server/protocols/completed_analysis_store.py]: 😆

File: robot-server/robot_server/protocols/rtp_resources.py [hunk 7]
```diff
@@ -71,3 +71,41 @@ def _bool_converter(value: str) -> bool:
             parameter_type=parameter_type.value,
             parameter_value=parameter_value,
         )
+
+
+@dataclass
+class CsvParameterResource:
```
Comment (jbleon95) [inline robot-server/robot_server/protocols/rtp_resources.py]: Sorry to be nitpicky on this but for style consistency lets also have this be `CSVParameterResource`

**Review by jbleon95 — APPROVED** [review]
Besides one more teensy nitpick this looks great.

**Pull request merged**